src/compiler/counted_array.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2022 Mozilla Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Partially sub-licensed under the MIT license
+// Copyright (c) 2016 Alex Burka
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+/**
+ * Helper macro to create fixed-length arrays without specifying a static size
+ */
+#[macro_export]
+macro_rules! counted_array {
+    // INTERNAL
+
+    // last element (proceed to output)
+    (@parse $size:expr, ($val:expr) -> [$($accs:expr),*] $thru:tt) => {
+        counted_array!(@output $size + 1usize, [$($accs,)* $val] $thru);
+    };
+    // more elements (keep parsing)
+    (@parse $size:expr, ($val:expr, $($vals:expr),*) -> [$($accs:expr),*] $thru:tt) => {
+        counted_array!(@parse $size + 1usize, ($($vals),*) -> [$($accs,)* $val] $thru);
+    };
+
+    // SCOPE
+    // output a static or const item
+    (@output $size:expr, $acc:tt (($($p:tt)*) $s:ident $n:ident $t:ty)) => {
+        $($p)* $s $n: [$t; $size] = $acc;
+    };
```

Comment:
```
macro_rules! countedarr {
    ($v:vis static $name:ident : [ $individual:ty ; _ ] = [$($arg:tt),*]; ) => {
        $v static $name : [$individual; countedarr!(count: $($arg,)*)] = [
            $($arg),*
        ];
    };
    (count: ) => { 0usize };
    (count: $head:tt, $($tail:tt,)*) => {
        1usize + countedarr!(count: $($tail,)*)
    };
}
```

this replaces the entire macro

Reply:
That doesn't work, it has the same issue as before. I spent more time trying to fix it, but I still wasn't able to get your version to work.
![image](https://user-images.githubusercontent.com/3370875/185743957-2cd49d3c-aae8-4b4c-b172-091c01817cdd.png)
![image](https://user-images.githubusercontent.com/3370875/185744256-c37c5928-af21-4333-86c7-2dafe0514081.png)

Reply:
Let me try it again, I just realized you aren't using the same name for the macro.

Reply:
Yes, I tried your macro, and I think I've figured out why it doesn't work (though this isn't definite).

The way that sccache is using that counted_array library currently involves expressions (such as more macros) and not tokens. Your approach is processing them as tokens, not expressions, which doesn't allow you to see what the actual underlying result of the wrapped macro will be. That is why the solution they used in `extern crate::counted_array` was to do recursion to evaluate the expressions into tokens. If you don't do that you can't have `counted_array!` macros containing other macros, a combination which we use extensively.

Reply:
You are correct, that was my bad using `tt` instead of `expr`. I went ahead and created #1296 since I didn't see you were continuing working on this. I'd appreciate your review on that so we can close this topic. Using expressions also resolves the usage with nested `macro_rules`.